File: NextPermutation/nextPermutation.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var nextPermutation = function(nums) {
  
  let flag = null;
  for (let i = nums.length - 1; i >= 1; i--) {
    if (nums[i - 1] < nums[i]) {
      flag = i - 1;
      break;
    }
  }

  if (flag != null) {
    for (let j = nums.length - 1; j > 0; j--) {
      if (nums[flag] < nums[j]) {
        [nums[flag], nums[j]] = [nums[j], nums[flag]];
        flag += 1;
        break;
      }
    }
    sort(nums, flag, nums.length - 1);
  } else {
    sort(nums, 0, nums.length - 1);
  }
  return;
};

function sort(nums, start, end) {
  for(let i = start; i < end; i++) {
    let min = i;
    for (let k = i + 1; k <= end; k++) {
      if (nums[min] > nums[k]) {
        min = k;
      }
    }
    [nums[i], nums[min]] = [nums[min], nums[i]];
  }
}

data = [1,3,2]

nextPermutation(data);
console.log(data);

File: NextPermutation/nextPermutation.py
class Solution():
    def nextPermutation(self, nums):

        sizeArr = len(nums) - 1
        index = -1
        for i in range(sizeArr, 0, -1):
            if nums[i] > nums[i - 1]:
                index = i - 1
                break

        if index == -1:
            nums.sort()
            return
        for i in range(sizeArr, index, -1):
            if (nums[i] > nums[index]):
                nums[i], nums[index] = nums[index], nums[i]
                break

        left = index + 1
        right = sizeArr
        while left < right:
            nums[left], nums[right] = nums[right], nums[left]
            left += 1
            right -= 1
        return
    
sol = Solution()
nums = [1,4,3,2]
sol.nextPermutation(nums)
print(nums)